src/ripple/beast/core/SharedPtr.h

```diff
@@ -176,7 +176,7 @@ class SharedPtr
     */
     /** @{ */
 #if BEAST_SHAREDPTR_PROVIDE_COMPILER_WORKAROUNDS
-    SharedPtr& operator= (SharedPtr && sp)
+    SharedPtr& operator= (SharedPtr && sp) noexcept
```

Comment:
This needs `noexcept` on `swap`.

Reply:
Right. And in making `swap` `noexcept`, we'll have to make several other functions `noexcept`. You eventually get to a point where we're imposing potentially new requirements on the type `U` for `acquire` and the type `T` for `release`. Namely, the functions `incReferenceCount` or `decReferenceCount` must be `noexcept`.

So, to continue down this rabbit hole: we need to analyze the usages in Ripple using `beast::SharedPtr`. One such example uses `Source` which inherits from `beast::SharedObject`. We would need to make its `decReferenceCount` `noexcept` which in turn means making `destroy` `noexcept`. I'd have to look at the other uses of `SharedPtr` to verify their ref count methods are `noexcept`.

It makes me wonder though: is this worth doing? It seems fragile as if someone uses `SharedPtr` with a new type and doesn't mark their ref count methods `noexcept`, I don't see an easy way to protect against that in the future.

Reply:
I was thinking (the member) `swap` just swaps pointers, side stepping that rabbit hole.  I missed the `assign` which takes us into that rabbit hole.  I know it's doable, because `std::shared_ptr` does it.  We should probably follow that model.  Otoh, perhaps the right thing to do is to ditch trying to fix `SharedPtr` and instead replace it with `std::shared_ptr` in a separate P/R instead.

Reply:
I'm with you on the latter idea. I think the right thing to do here is to just replace usages with `std::shared_ptr` and remove this dependency.  I'll create a GitHub issue for that and likely work on it next week after the holiday.

Reply:
Ended up just doing this now -- PR is https://github.com/ripple/rippled/pull/2557